### DISCUSSION TITLE
Git config nonstandard location breaks the Git dependent toolchain

### DISCUSSION THREAD

**Discussion by shcheklein** [community/community#38070] [Codespaces] [answered]
It seems Codespaces is using a custom Git build that puts its config into a nonstandard location `/usr/local/etc/gitconfig` to set user name, email, etc:

```bash
@shcheklein ➜ /workspaces/example-get-started (main) $ /usr/local/bin/git config --list --show-origin
file:/usr/local/etc/gitconfig   credential.helper=/.codespaces/bin/gitcredential_github.sh
file:/usr/local/etc/gitconfig   user.name=Ivan Shcheklein
file:/usr/local/etc/gitconfig   user.email=<email>
file:/usr/local/etc/gitconfig   gpg.program=/.codespaces/bin/gh-gpgsign
```

Libraries like `dulwich`, `pygit2`, and all the code that depends on them (and probably there are other examples in the other languages) are broken as a result of this. It means for example, that [DVC](https://github.com/iterative/dvc) is also broken, means that DVC extension is also broken ... you got the idea.

GitLab/Gitpod are using the standard location `/etc/gitconfig`. I think Codespaces was using the standard location as well before.

What was the core reason behind this change?
Can we find a better / standard way that doesn't break the toolchain?

**Comment by pmrowla** [comment]
To add to @shcheklein's post, this also affects regular CLI git (and not only libraries like libgit2 or dulwich) if you use any `git` besides what I assume is the one installed by github at `/usr/local/bin/git`.

In the codespaces container, by default your `$PATH` will prefer `/usr/local/bin/git`, but if a user overrides `$PATH` (like via a configured dotfiles repo) or just runs a script that calls `/usr/bin/git` directly, it will also not respect `/usr/local/etc/gitconfig`

Example using the standard microsoft devcontainers `python` container (which includes a `git` installed in `/usr/bin/git`):

No `user.name` available due to `/usr/bin/git` checking nonexistent `/etc/gitconfig`:
```
@pmrowla ➜ /workspaces/example-get-started (main) $ /usr/bin/git config --list --show-origin
file:.git/config        core.repositoryformatversion=0
file:.git/config        core.filemode=true
file:.git/config        core.bare=false
file:.git/config        core.logallrefupdates=true
file:.git/config        remote.origin.url=https://github.com/iterative/example-get-started
file:.git/config        remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config        branch.main.remote=origin
```

`user.name` available due to `/usr/local/bin/git` checking `/usr/local/etc/gitconfig`:
```
@pmrowla ➜ /workspaces/example-get-started (main) $ /usr/local/bin/git config --list --show-origin
file:/usr/local/etc/gitconfig   credential.helper=/.codespaces/bin/gitcredential_github.sh
file:/usr/local/etc/gitconfig   user.name=Peter Rowlands (변기호)
file:/usr/local/etc/gitconfig   user.email=peter@pmrowla.com
file:/usr/local/etc/gitconfig   gpg.program=/.codespaces/bin/gh-gpgsign
file:.git/config        core.repositoryformatversion=0
file:.git/config        core.filemode=true
file:.git/config        core.bare=false
file:.git/config        core.logallrefupdates=true
file:.git/config        remote.origin.url=https://github.com/iterative/example-get-started
file:.git/config        remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config        branch.main.remote=origin
file:.git/config        branch.main.merge=refs/heads/main
```

I can understand why codespaces includes a git installed in `/usr/local` (since there's no guarantee that the user's container includes `git` or includes a `git` version that's too old), but in my opinion codespaces should still place a system gitconfig in the standard location (`/etc/gitconfig`) to ensure that it is loaded regardless of whether or not the user is using github's `git` installation.

This could just be done with something as simple as the codespaces container adding a symlink from `/etc/gitconfig -> /usr/local/etc/gitconfig` by default.

**Reply by david-engelmann** [reply]
`adding a symlink from /etc/gitconfig -> /usr/local/etc/gitconfig by default`

+1

**Comment by reybard** [comment, marked as answer]
Thank you for the report.

I have seen some internal messaging indicating this has been a problem for our own developers as well, so I've created a ticket on our board to track. I don't see a reason why the symlink could not be added, but we are currently in a code freeze leading up to Universe so it may be a few days before a fix is pushed out.

**Reply by shcheklein** [reply]
@reybard thanks, makes sense. As a temporary workaround, I'm considering to do something like https://github.com/iterative/features/blob/main/src/dvc/install.sh#L15-L20 in the feature that we are using in the conference presentation + in all examples. Would that work, or features are setup before this custom Git is installed?

**Reply by shcheklein** [reply]
okay, this won't work (I think git config is created later in the process for this to work). And `postCreateCommand` is also not supported. Not sure if there is a way to do a workaround on the feature level

**Comment by jkeech** [comment]
I suspect this was an unintended side effect of switching the images to [build git from source](https://github.com/devcontainers/images/pull/162) a couple weeks ago in order to quickly pick up the latest Git versions from https://github.blog/2022-10-18-git-security-vulnerabilities-announced/ before those patched `git` builds were available on upstream apt repositories. We should be able to switch this back to the original paths without needing to add a new symlink.
